src/vm-vcpu/src/vcpu/mod.rs

```diff
@@ -244,81 +245,91 @@ impl KvmVcpu {
                         }
-                        unsafe { libc::exit(0) };
-                    }
-                    VcpuExit::IoOut(addr, data) => {
-                        if 0x3f8 <= addr && addr < (0x3f8 + 8) {
-                            // Write at the serial port.
+                        VcpuExit::IoOut(addr, data) => {
+                            if 0x3f8 <= addr && addr < (0x3f8 + 8) {
+                                // Write at the serial port.
+                                if self
+                                    .device_mgr
+                                    .lock()
+                                    .unwrap()
+                                    .pio_write(PioAddress(addr), data)
+                                    .is_err()
+                                {
+                                    eprintln!("Failed to write to serial port");
+                                }
+                            } else if addr == 0x060 || addr == 0x061 || addr == 0x064 {
+                                // Write at the i8042 port.
+                                // See https://wiki.osdev.org/%228042%22_PS/2_Controller#PS.2F2_Controller_IO_Ports
+                            } else if 0x070 <= addr && addr <= 0x07f {
+                                // Write at the RTC port.
+                            } else {
+                                // Write at some other port.
+                            }
+                        }
+                        VcpuExit::IoIn(addr, data) => {
+                            if 0x3f8 <= addr && addr < (0x3f8 + 8) {
+                                // Read from the serial port.
+                                if self
+                                    .device_mgr
+                                    .lock()
+                                    .unwrap()
+                                    .pio_read(PioAddress(addr), data)
+                                    .is_err()
+                                {
+                                    eprintln!("Failed to read from serial port");
+                                }
+                            } else {
+                                // Read from some other port.
+                            }
+                        }
+                        VcpuExit::MmioRead(addr, data) => {
                             if self
                                 .device_mgr
                                 .lock()
                                 .unwrap()
-                                .pio_write(PioAddress(addr), data)
+                                .mmio_read(MmioAddress(addr), data)
                                 .is_err()
                             {
-                                eprintln!("Failed to write to serial port");
+                                eprintln!("Failed to read from mmio");
                             }
-                        } else if addr == 0x060 || addr == 0x061 || addr == 0x064 {
```

Comment:
I'd keep these around even though they're not called yet; it still exemplifies what to do for i8042 exits, and we'll get to them anyway soon.

Reply:
They're still there, but GitHub is showing the diff a bit weirdly. They're at line 323 now.